File: native/engine/core/math/math.cpp
#include "math.h"
#include <stdlib.h>

#ifdef __EMSCRIPTEN__
#include <emscripten.h>
#endif

/**
 * 角度转弧度
 */
float Math::DEG_TO_RAD = Math::PI / 180;

/**
 * 弧度转角度
 */
float Math::RAD_TO_DEG = 180 / Math::PI;

float Math::getAngle(float x, float y)
{
	/**
	 * atan2返回与x轴的夹角弧度
	 *            90
	 * 逆时针      |  逆时针
	 * [90,180]   |  [0,90]
	 * -------------------------
	 * [-90,-180] |  [0,-90]
	 * 顺时针      |  顺时针
	 *            -90
	 */
	float r = ::atan2(y, x);
	r *= Math::RAD_TO_DEG;
	return r < 0 ? 360 + r : r;
}

bool Math::equlas(float a, float b, float epsilon)
{
	return fabs(a - b) < epsilon;
}

float Math::angleToRadian(float angle)
{
	return angle * Math::DEG_TO_RAD;
}

float Math::radianToAngle(float radian)
{
	return radian * Math::RAD_TO_DEG;
}

// 生成 [0.0, 1.0) 的随机浮点数
double random_float_0_1()
{
	return (double)rand() / (RAND_MAX + 1.0);
}

float Math::randomFloat(float min, float max)
{
#ifdef __EMSCRIPTEN__
	return min + emscripten_random() * max;
#else
	return min + (max - min) * random_float_0_1();
#endif
}

float Math::lerp(float a, float b, float alpha)
{
	return a + (b - a) * alpha;
}

double Math::fmod(double p_x, double p_y)
{
	return ::fmod(p_x, p_y);
}
float Math::fmod(float p_x, float p_y)
{
	return ::fmodf(p_x, p_y);
}

float Math::ceil(float v)
{
	return ::ceil(v);
}

float Math::abs(float v)
{
	return ::abs(v);
}

int Math::abs(int v)
{
	return ::abs(v);
}

float Math::floor(float v)
{
	return ::floor(v);
}

float Math::asin(float angle)
{
	return ::asin(angle);
}

float Math::cos(float angle)
{
	return ::cos(angle);
}

float Math::sin(float angle)
{
	return ::sin(angle);
}

float Math::atan2(float x, float y)
{
	return ::atan2(y, x);
}

#if defined(_MSC_VER) && !defined(__clang__)
#include <intrin.h>

// https://en.wikipedia.org/wiki/Find_first_set

uint32_t Math::CTZ32(uint32_t block)
{
	unsigned long index;
	_BitScanForward(&index, block);
	return index;
}

// This function doesn't need to be fast, so using the Ivy Bridge fallback.
uint32_t Math::CLZ32(uint32_t value)
{
#if 1
	// Use BSR (Bit Scan Reverse) which is available on Ivy Bridge
	unsigned long index;
	if (_BitScanReverse(&index, value))
	{
		// BSR gives the index of the most significant 1-bit
		// We need to invert this to get the number of leading zeros
		return 31 - index;
	}
	else
	{
		// If x is 0, BSR sets the zero flag and doesn't modify index
		// LZCNT should return 32 for an input of 0
		return 32;
	}

#else
	return __lzcnt(value);

#endif
}

uint32_t Math::CTZ64(uint64_t block)
{
	unsigned long index;

#ifdef _WIN64
	_BitScanForward64(&index, block);
#else
	// 32-bit fall back
	if ((uint32_t)block != 0)
	{
		_BitScanForward(&index, (uint32_t)block);
	}
	else
	{
		_BitScanForward(&index, (uint32_t)(block >> 32));
		index += 32;
	}
#endif

	return index;
}

#else

uint32_t Math::CTZ32(uint32_t block)
{
	return __builtin_ctz(block);
}

uint32_t Math::CLZ32(uint32_t value)
{
	return __builtin_clz(value);
}

uint32_t Math::CTZ64(uint64_t block)
{
	return __builtin_ctzll(block);
}

#endif

File: sq/core/math/math.ts

export default {


    /**
     * 角度转弧度
     */
    DEG_TO_RAD: Math.PI / 180,

    /**
     * 弧度转角度
     */
    RAD_TO_DEG: 180 / Math.PI,

    /**
     * 角度转弧度
     */
    degreesToRadians(angle: number) {
        return angle * this.RAD_TO_DEG;
    },

    clamp(value: number, min: number, max: number) {
        if (value >= max) return max;
        if (value <= min) return min;
        return value;
    },


    /**
     * 将值限制在0~1之间
     * @param value 
     */
    clamp01(value: number): number {
        return this.clamp(value, 0, 1);
    },

    lerp(a: number, b: number, alpha: number) {
        return a + (b - a) * this.clamp(alpha, 0, 1);
    },

    lerpAngle(a: number, b: number, alpha: number) {
        if (b - a > 180) {
            b -= 360;
        }
        if (b - a < -180) {
            b += 360;
        }
        return this.lerp(a, b, this.clamp(alpha, 0, 1));
    },


    powerOfTwo(x: number) {
        return ((x !== 0) && !(x & (x - 1)));
    },

    nextPowerOfTwo(val: number) {
        val--;
        val |= (val >> 1);
        val |= (val >> 2);
        val |= (val >> 4);
        val |= (val >> 8);
        val |= (val >> 16);
        val++;
        return val;
    },

    random(min: number, max: number) {
        var diff = max - min;
        return Math.random() * diff + min;
    },


    smoothstep(min: number, max: number, x: number) {
        if (x <= min) return 0;
        if (x >= max) return 1;

        x = (x - min) / (max - min);

        return x * x * (3 - 2 * x);
    },

    smootherstep(min: number, max: number, x: number) {
        if (x <= min) return 0;
        if (x >= max) return 1;

        x = (x - min) / (max - min);

        return x * x * x * (x * (x * 6 - 15) + 10);
    },

    fsel(a: number, b: number, c: number) {
        return a >= 0 ? b : c;
    },


    /**
     * 分段插值函数，返回[1,-1]范围的值
     * t <= 0.5 返回[0,1]
     * t>=0.5 返回 (0,-1]
     * @param t 
     * @returns 
     */
    smootherUpDown(t: number) {
        if (t <= 0.5)
            return 2 * t;
        else
            return 2 - 2 * t;
    },

    /**
     * 返回向量在一个圆周内的角度，与x轴的夹角
     * 以逆时针方向为准，返回0～360范围的角度。如果向量与x轴重叠即(1,0)返回0度，（-1，0）返回180度
     */
    getAngle(x: number, y: number) {
        let r: number = Math.atan2(y, x);
        r *= this.RAD_TO_DEG;
        return r < 0 ? 360 + r : r;
    },

    /**
     * 第二象限（90° ~ 180°） → 转换到 0° ~ 90°
     * 第一象限（0° ~ 90°） → 转换到 270° ~ 360
     * 第三象限（180° ~ 270°） → 转换到 90° ~ 180°
     * 第四象限（270° ~ 360°） → 转换到 180° ~ 270°
     * 
     * @param x 
     * @param y 
     * @returns 
     */
    convertRotateAngle(x: number, y: number): number {
        let angle = Math.atan2(y, x) * this.RAD_TO_DEG;
        angle = (angle + 360) % 360; // 转换到 [0, 360] 范围

        // 使第二象限为 0°，逆时针旋转
        // angle = (angle - 90 + 360) % 360;
        angle = (angle + 270) % 360;
        return angle;
    },

    getReflectedAngleX(x: number, y: number): number {
        // 计算反射向量 (-x, y)
        return this.convertRotateAngle(-x, y);
    },

    equlas(a: number, b: number, epsilon: number = 0.1): boolean {
        return Math.abs(a - b) < epsilon;
    },

    /**
     *  返回t / length的浮点余数。
     * @param t 
     * @param length 
     * @returns 
     */
    repeat(t: number, length: number): number {
        return t - Math.floor(t / length) * length;
    }
} as {
    DEG_TO_RAD: number;
    RAD_TO_DEG: number;
    degreesToRadians(angle: number): number;
    clamp(value: number, min: number, max: number): number;
    clamp01(value: number): number;
    lerp(a: number, b: number, alpha: number): number;
    getAngle(x: number, y: number): number;
    smootherUpDown(t: number): number;
    convertRotateAngle(x: number, y: number): number;
    getReflectedAngleX(x: number, y: number): number;
    equlas(a: number, b: number, epsilon?: number): boolean;
    repeat(t: number, length: number): number;
    random(min: number, max: number): number;
}
